pkg/client/kubernetes/client.go

```diff
@@ -382,15 +383,33 @@ var _ client.Client = &FallbackClient{}
 // in case Get/List requests with the ordinary `client.Client` fail (e.g. because of cache errors).
 type FallbackClient struct {
 	client.Client
-	Reader client.Reader
+	Reader           client.Reader
+	KindToNamespaces map[string]sets.Set[string]
 }
 
 var cacheError = &kubernetescache.CacheError{}
 
 // Get retrieves an obj for a given object key from the Kubernetes Cluster.
 // In case of a cache error, the underlying API reader is used to execute the request again.
 func (d *FallbackClient) Get(ctx context.Context, key client.ObjectKey, obj client.Object, opts ...client.GetOption) error {
-	err := d.Client.Get(ctx, key, obj)
+	gvk, err := apiutil.GVKForObject(obj, GardenScheme)
+	if err != nil {
+		return err
+	}
+
+	// Check if there are specific namespaces for this object's kind in the cache.
+	namespaces, ok := d.KindToNamespaces[gvk.Kind]
+
+	// If there are specific namespaces for this kind in the cache and the object's namespace is not cached,
+	// use the API reader to get the object.
+	if ok && !namespaces.Has(obj.GetNamespace()) {
```

Comment:
Shouldn't it be like this?

```suggestion
	if !ok || !namespaces.Has(obj.GetNamespace()) {
```

Reply:
No. Only if for a kind, a list of namespace is present and that list of namespace doesn't have the namespace for the current object then only we want to go to the reader in all other cases first try always from the cache client.

Reply:
Makes sense, thank you!